### PR TITLE
General improvements

### LINKED ISSUE #268: Add ability for "named" addresses
Other explorers will display a name/tag for known addresses, such as an exchange's withdrawal address.

We are considering removing explorer.go and merging the API into the skycon node directly, there isn't a need for the explorer.go proxy any more, because we have the `-enable-api-sets` feature in Skycoin core.

In that case, this annotation data would have to be stored in the Skycoin node.  And we would need a mechanism for annotating these addresses (such as an admin control).  Initially we could load from a JSON file on disk.

This may tie into a general annotation/notes system in Skycoin core.

### LINKED ISSUE #293: Use the node API directly
Currently the explorer uses an intermediate server to obtain the data, but there isn't a need for the explorer.go proxy any more, because we have the `-enable-api-sets` feature in Skycoin core.

@gz-c, do you think there is something that should to be done before starting this change?

### PR DESCRIPTION
**Pull request by Senyoret1**
Fixes #268
Fixes #293

Changes:
- Updates Angular from v8 to v11.
- Adds a way to assign labels to specific addresses.
- Adds a mode for using the explorer without the intermediate Go server. It is disabled by default.
- The error messages shown when an adderess is invalid and when the address in empty were improved.
- The outputs page now shows partial results in some circunstances, to improve the performance, just as the address page does.
- Unused test files were removed.
- Some other small improvements.

Note: moved from the previous repo.

Does this change need to mentioned in CHANGELOG.md?
No

### CODE REVIEW

**Comment by Senyoret1** [comment]
The Travis build will work after #353 is merged.

I did not found problems with Snyk, so I will try again after #353 is merged too.

**Comment by Senyoret1** [comment]
The last commit updates the dependencies to make the Snyk tests pass.

File: src/app/components/pages/blocks/blocks.component.ts [hunk 1]
```diff
@@ -85,7 +90,12 @@ export class BlocksComponent implements OnInit, OnDestroy {
     private api: ApiService,
     private explorer: ExplorerService,
     private route: ActivatedRoute,
-  ) { }
+  ) {
+    // Each time the node URL is changed, check if a local node is being used.
+    this.nodeUrlSubscription = this.api.localNodeUrl.subscribe(response => {
```
Comment (SkycoinSynth) [inline src/app/components/pages/blocks/blocks.component.ts:95]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    this.nodeUrlSubscription = this.api.localNodeUrl.subscribe((response) => {
```

File: src/app/components/pages/node-url.component.ts [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import { Subscription } from 'rxjs';
+import { Component, OnInit, OnDestroy } from '@angular/core';
```
Comment (SkycoinSynth) [inline src/app/components/pages/node-url.component.ts:2]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
import { Component, OnInit, OnDestroy } from "@angular/core";
```

File: src/app/services/explorer/explorer.service.ts [hunk 3]
```diff
@@ -224,4 +240,16 @@ export class ExplorerService {
...
       map(response => parseGetTransaction(response)));
   }
+
+  /**
+   * Returns the name associated to an address, enclosed in parentheses. If no name has been
+   * associated to the address, returns an empty string.
+   */
+  getAddressName(address: string): string {
+    if (this.namedAddressesMap.has(address)) {
+      return ' (' + this.namedAddressesMap.get(address) + ')';
```
Comment (SkycoinSynth) [inline src/app/services/explorer/explorer.service.ts:250]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
      return " (" + this.namedAddressesMap.get(address) + ')';
```

File: src/app/components/pages/node-url.component.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+import { ActivatedRoute, Router } from '@angular/router';
+
+import { ApiService } from 'app/services/api/api.service';
```
Comment (SkycoinSynth) [inline src/app/components/pages/node-url.component.ts:5]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
import { ApiService } from "app/services/api/api.service";
```

File: src/app/app.component.spec.ts [hunk 5]
```diff
@@ -1,32 +1,40 @@
-import { TestBed, async } from '@angular/core/testing';
+import { TestBed, waitForAsync } from '@angular/core/testing';
+import { Router } from '@angular/router';
+import { of } from 'rxjs';
 
 import { AppComponent } from './app.component';
+import { LanguageService } from './services/language/language.service';
+import { ExplorerService } from './services/explorer/explorer.service';
```
Comment (SkycoinSynth) [inline src/app/app.component.spec.ts:7]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
import { ExplorerService } from "./services/explorer/explorer.service";
```

File: src/app/app.config.ts [hunk 6]
```diff
@@ -95,3 +95,13 @@ export const languageConfig = {
    */
   defaultLanguage: 'en'
 };
+
+/**
+ * List with named addresses.
+ */
+export const namedAddresses = [
+  {
+    address: '2iNNt6fm9LszSWe51693BeyNUKX34pPaLx8',
```
Comment (SkycoinSynth) [inline src/app/app.config.ts:104]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    address: "2iNNt6fm9LszSWe51693BeyNUKX34pPaLx8",
```

File: src/app/services/api/api.service.ts [hunk 7]
```diff
@@ -98,38 +158,57 @@ export class ApiService {
...
+    return this.get(url, { addrs: address });
   }
 
   /**
    * Gets a transaction by its hash.
    * @param transactionId Transaction hash.
    */
   getTransaction(transactionId: string): Observable<GetTransactionResponse> {
-    return this.get('transaction', { txid: transactionId, verbose: 1 });
+    const url = !this.nodeUrl() ? 'transaction' : 'v1/transaction';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:191]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "transaction" : 'v1/transaction';
```

File: src/app/components/pages/address-detail/address-detail.component.ts [hunk 8]
```diff
@@ -289,12 +289,20 @@ export class AddressDetailComponent implements OnInit, OnDestroy {
...
+        this.alltransactions = undefined;
+      }
+
       this.dataLoaded = true;
     }, error => {
       if (error.status >= 400 && error.status < 500) {
         // The address was not found.
         this.loadingMsg = 'general.noData';
-        this.longErrorMsg = 'addressDetail.withoutTransactions';
+        this.longErrorMsg = 'addressDetail.invalidAddress';
```
Comment (SkycoinSynth) [inline src/app/components/pages/address-detail/address-detail.component.ts:305]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
        this.longErrorMsg = "addressDetail.invalidAddress";
```

File: src/app/components/layout/search-bar/search-bar.component.ts [hunk 9]
```diff
@@ -1,9 +1,11 @@
 import { Component, ElementRef, ViewChild, OnDestroy } from '@angular/core';
 import { Router } from '@angular/router';
 import { TranslateService } from '@ngx-translate/core';
-import { Subscription } from 'rxjs';
+import { Subscription, of } from 'rxjs';
+import { delay, mergeMap } from 'rxjs/operators';
```
Comment (SkycoinSynth) [inline src/app/components/layout/search-bar/search-bar.component.ts:5]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
import { delay, mergeMap } from "rxjs/operators";
```

File: src/app/components/pages/node-url.component.ts [hunk 10]
```diff
@@ -0,0 +1,52 @@
+import { Subscription } from 'rxjs';
```
Comment (SkycoinSynth) [inline src/app/components/pages/node-url.component.ts:1]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
import { Subscription } from "rxjs";
```

File: src/app/services/explorer/explorer.service.ts [hunk 11]
```diff
@@ -54,24 +57,34 @@ export class ExplorerService {
...
 
+  /**
+   * Initializes the service.
+   */
   initialize() {
-    if (this.initialized) {
-      return;
-    }
-    this.initialized = true;
+    namedAddresses.forEach(namedAddress => {
```
Comment (SkycoinSynth) [inline src/app/services/explorer/explorer.service.ts:71]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    namedAddresses.forEach((namedAddress) => {
```

File: src/app/services/api/api.service.ts [hunk 12]
```diff
@@ -98,38 +158,57 @@ export class ApiService {
...
+    const url = !this.nodeUrl() ? 'richlist' : 'v1/richlist';
+
+    return this.get(url).pipe(map((response: any) => response.richlist));
+  }
+
+  /**
+   * Gets the sync state of the node.
+   */
+  getSyncState(): Observable<GetSyncStateResponse> {
+    const url = !this.nodeUrl() ? 'blockchain/progress' : 'v1/blockchain/progress';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:209]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "blockchain/progress" : 'v1/blockchain/progress';
```

File: src/app/services/api/api.service.ts [hunk 13]
```diff
@@ -98,38 +158,57 @@ export class ApiService {
...
+    return this.get(url, { addrs: address });
   }
 
   /**
    * Gets the balance of an address.
    * @param address Address to consult.
    */
   getBalance(address: string): Observable<GetBalanceResponse> {
-    return this.get('balance', { addrs: address });
+    const url = !this.nodeUrl() ? 'balance' : 'v1/balance';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:181]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "balance" : 'v1/balance';
```

File: src/app/components/layout/search-bar/search-bar.component.ts [hunk 14]
```diff
@@ -24,21 +26,68 @@ export class SearchBarComponent implements OnDestroy {
...
+   * repeated after a delay.
+   * @param delayMs Delay before performing the actual check.
+   */
+  private checkSyncState(delayMs: number) {
+    if (this.syncSubscription) {
+      this.syncSubscription.unsubscribe();
+    }
+
+    // Get the data.
+    this.syncSubscription = of(0).pipe(delay(delayMs), mergeMap(() => this.apiService.getSyncState())).subscribe(response => {
```
Comment (SkycoinSynth) [inline src/app/components/layout/search-bar/search-bar.component.ts:73]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    this.syncSubscription = of(0).pipe(delay(delayMs), mergeMap(() => this.apiService.getSyncState())).subscribe((response) => {
```

File: src/app/services/api/api.service.ts [hunk 15]
```diff
@@ -98,38 +158,57 @@ export class ApiService {
...
+
+    return this.get(url, { txid: transactionId, verbose: 1 });
   }
 
   /**
    * Gets the list of unlocked addresses with most coins.
    */
   getRichlist(): Observable<RichlistEntry[]> {
-    return this.get('richlist').pipe(map((response: any) => response.richlist));
+    const url = !this.nodeUrl() ? 'richlist' : 'v1/richlist';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:200]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "richlist" : 'v1/richlist';
```

File: src/app/components/pages/node-url.component.ts [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    private apiService: ApiService,
+  ) { }
+
+  ngOnInit() {
+    // Check the URL params to detect the URL of the local node.
+    this.navParamsSubscription = this.route.params.subscribe(params => {
+      let nodeUrl: string = params['url'];
+
+      // If the url is "null", convert the value to a real null.
+      if (nodeUrl.toUpperCase() === 'null'.toUpperCase()) {
```
Comment (SkycoinSynth) [inline src/app/components/pages/node-url.component.ts:35]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
      if (nodeUrl.toUpperCase() === "null".toUpperCase()) {
```

File: src/app/components/pages/node-url.component.ts [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * Page for setting the URL of the local node the app must use as backend. The new URL is saved
+ * in local storage, so it is used again the next time the app is openned using the same
+ * domain name. If the URL is the string "null", any previously saved URL is deleted. Just after
+ * saving the URL, the browser is redirected to the main page.
+ */
+@Component({
+  selector: 'app-node-url',
+  template: '',
```
Comment (SkycoinSynth) [inline src/app/components/pages/node-url.component.ts:16]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
  template: "",
```

File: src/app/services/api/api.service.ts [hunk 18]
```diff
@@ -81,14 +137,18 @@ export class ApiService {
...
+
+    return this.get(url, { start: startNumber, end: endNumber });
   }
 
   /**
    * Gets information about the state of the blockchain.
    */
   getBlockchainMetadata(): Observable<Blockchain> {
-    return this.get('blockchain/metadata').pipe(
+    const url = !this.nodeUrl() ? 'blockchain/metadata' : 'v1/blockchain/metadata';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:149]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "blockchain/metadata" : 'v1/blockchain/metadata';
```

File: e2e/address-detail/address-detail.po.ts [hunk 19]
```diff
@@ -8,7 +8,7 @@ export class AddressDetailPage {
 
   getAddressForSmallScreens() {
     return element(by.css('.element-details > div:nth-of-type(1) > span:nth-of-type(2)'))
-      .getAttribute('textContent');
+      .getAttribute('textContent').then(address => address.trim());
```
Comment (SkycoinSynth) [inline e2e/address-detail/address-detail.po.ts:11]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
      .getAttribute("textContent").then(address => address.trim());
```

File: src/app/services/api/api.service.ts [hunk 20]
```diff
@@ -18,61 +18,117 @@ import { Blockchain, GetBlocksResponse, GetBlockchainMetadataResponse, GetUnconf
...
+  get localNodeUrl(): Observable<string> {
+    return this.localNodeUrlSubject.asObservable();
+  }
+
   /**
    * Get information about the state of the node.
    */
   getHealth(): Observable<any> {
-    return this.get('health');
+    const url = !this.nodeUrl() ? 'health' : 'v1/health';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:79]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "health" : 'v1/health';
```

File: src/app/services/api/api.service.ts [hunk 21]
```diff
@@ -18,61 +18,117 @@ import { Blockchain, GetBlocksResponse, GetBlockchainMetadataResponse, GetUnconf
...
+    return this.get(url);
   }
 
   /**
    * Gets the list of transactions of a specific address.
    * @param address Address to consult.
    */
   getAddress(address: string): Observable<GetTransactionResponse[]> {
-    return this.get('transactions', { addrs: address, verbose: 1 });
+    const url = !this.nodeUrl() ? 'transactions' : 'v1/transactions';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:89]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "transactions" : 'v1/transactions';
```

File: src/app/services/api/api.service.ts [hunk 22]
```diff
@@ -18,61 +18,117 @@ import { Blockchain, GetBlocksResponse, GetBlockchainMetadataResponse, GetUnconf
...
+    return this.get(url, { addrs: address, verbose: 1 });
   }
 
   /**
    * Gets the list of transactions of a specific address, using pagination.
    * @param address Address to consult.
    */
   getAddressWithPagination(address: string, page: number, pageSize: number): Observable<any> {
-    return this.get('paginatedTransactions', { addrs: address, page: page, limit: pageSize, sort: 'desc', verbose: 1 })
+    const url = !this.nodeUrl() ? 'paginatedTransactions' : 'v2/transactions';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:99]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "paginatedTransactions" : 'v2/transactions';
```

File: src/app/services/api/api.service.ts [hunk 23]
```diff
@@ -18,61 +18,117 @@ import { Blockchain, GetBlocksResponse, GetBlockchainMetadataResponse, GetUnconf
...
+    return this.get(url, { verbose: 1 });
   }
 
   /**
    * Gets a block by its ID (sequence number).
    * @param id Block ID (sequence number).
    */
   getBlockById(id: number): Observable<GenericBlockResponse> {
-    return this.get('block', { seq: id, verbose: 1 });
+    const url = !this.nodeUrl() ? 'block' : 'v1/block';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:119]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "block" : 'v1/block';
```

File: src/app/services/explorer/explorer.service.ts [hunk 24]
```diff
@@ -54,24 +57,34 @@ export class ExplorerService {
...
+  /**
+   * Gets the basic info about the backend.
+   * @param delayMs Delay before starting to get the data.
+   */
+  private getNodeInfo(delayMs: number) {
+    if (this.initializationSubscription) {
+      this.initializationSubscription.unsubscribe();
+    }
 
+    this.initializationSubscription = of(0).pipe(delay(delayMs), mergeMap(() => this.api.getHealth())).subscribe(response => {
```
Comment (SkycoinSynth) [inline src/app/services/explorer/explorer.service.ts:87]: Codacy has a fix for the issue: [Expected parentheses around arrow function argument.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    this.initializationSubscription = of(0).pipe(delay(delayMs), mergeMap(() => this.api.getHealth())).subscribe((response) => {
```

File: src/app/services/api/api.service.ts [hunk 25]
```diff
@@ -98,38 +158,57 @@ export class ApiService {
...
+    return this.get(url);
   }
 
   /**
    * Gets the list of unspent outputs of an address.
    * @param address Address to consult.
    */
   getCurrentBalance(address: string): Observable<GetCurrentBalanceResponse> {
-    return this.get('currentBalance', { addrs: address });
+    const url = !this.nodeUrl() ? 'currentBalance' : 'v1/outputs';
```
Comment (SkycoinSynth) [inline src/app/services/api/api.service.ts:171]: Codacy has a fix for the issue: [Strings must use doublequote.](https://app.codacy.com/gh/skycoin/skycoin-explorer/pullRequest?prid=7163626)
```suggestion
    const url = !this.nodeUrl() ? "currentBalance" : 'v1/outputs';
```

**Review by iketheadore — APPROVED** [review]

**Pull request merged**